File: lab2/gr8sub2/Пардаев_Сергей/src/fork_example.py
import os
import time


def child_process(name):
    print(f"{name}: PID={os.getpid()}, PPID={os.getppid()}", flush=True)
    time.sleep(1)
    os._exit(0)


def main():
    print(f"parent: PID={os.getpid()}", flush=True)

    try:
        pid1 = os.fork()
    except OSError as e:
        print(f"Fork failed for child1: {e}", flush=True)
        pid1 = -1

    if pid1 == 0:
        child_process("child1")

    try:
        pid2 = os.fork()
    except OSError as e:
        print(f"Fork failed for child2: {e}", flush=True)
        pid2 = -1

    if pid2 == 0:
        child_process("child2")

    if pid1 > 0:
        os.waitpid(pid1, 0)
    if pid2 > 0:
        os.waitpid(pid2, 0)

    print("parent: children exited", flush=True)


if __name__ == "__main__":
    main()

File: lab2/gr9-sub1/RozhdaevEV/src/fork_example.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    pid_t pid1, pid2;
    int status1, status2;

    printf("parent: PID=%d, PPID=%d\n", getpid(), getppid());
    fflush(stdout);

   
    pid1 = fork();
    if (pid1 == 0) {
     
        printf("child_A: PID=%d, PPID=%d\n", getpid(), getppid());
        fflush(stdout);
        sleep(2);
        exit(0);
    } else if (pid1 > 0) {
       
       
        pid2 = fork();
        if (pid2 == 0) {
          
            printf("child_B: PID=%d, PPID=%d\n", getpid(), getppid());
            fflush(stdout);
            sleep(1);
            exit(0);
        } else if (pid2 > 0) {
            printf("parent: waiting for children...\n");
            fflush(stdout);
            
            waitpid(pid1, &status1, 0);
            printf("parent: child_A (PID=%d) finished with status %d\n", pid1, WEXITSTATUS(status1));
            fflush(stdout);
            
            waitpid(pid2, &status2, 0);
            printf("parent: child_B (PID=%d) finished with status %d\n", pid2, WEXITSTATUS(status2));
            fflush(stdout);
            
            printf("parent: all children finished\n");
            fflush(stdout);
        } else {
            perror("fork failed for child_B");
            exit(1);
        }
    } else {
        perror("fork failed for child_A");
        exit(1);
    }

    return 0;
}
